### ISSUE TITLE
specific-constant ConfDir is inconsistent in icinga2 r2.10.2-1

### ISSUE THREAD

**Issue by abaumgaertner** [Icinga/icinga2#6993]
icinga2 only partially respects -D ConfDir

ZonesDir | Read-only. Contains the path of the zones.d directory. Defaults to ConfigDir + "/zones.d".
--> stays at /etc/icinga2/zones.d when ConfigDir is set.

https://icinga.com/docs/icinga2/latest/doc/17-language-reference/#icinga-2-specific-constants

## Expected Behavior

when ConfDir is set, all conf files including icing2.conf, constants.conf, etc. and feature related files should be read from the specified directory
see output below and with troubleshoot option

## Current Behavior

systemd service running as icinga2
/usr/sbin/icinga2 daemon --log-level debug -D ConfDir=/appl/icinga/icinga2/etc/icinga2 -D LogDir=/appl/logs/icinga/icinga2 -D DataDir=/appl/icinga/icinga2/var/data -D CacheDir=/appl/icinga/icinga2/var/cache -D SpoolDir=/appl/icinga/icinga2/var/spool -D InitRunDir=/appl/icinga/icinga2/var/run

[root@avl4119p icingaweb2]# icinga2 api setup -D ConfDir=/appl/icinga/icinga2/etc/icinga2 -D LogDir=/appl/logs/icinga/icinga2 -D DataDir=/appl/icinga/icinga2/var/data -D CacheDir=/appl/icinga/icinga2/var/cache -D SpoolDir=/appl/icinga/icinga2/var/spool -D InitRunDir=/appl/icinga/icinga2/var/run
information/cli: Generating new CA.
information/base: Writing private key to '/appl/icinga/icinga2/var/data/ca//ca.key'.
information/base: Writing X509 certificate to '/appl/icinga/icinga2/var/data/ca//ca.crt'.
information/cli: Generating new CSR in '/appl/icinga/icinga2/var/data/certs//avl4119p.it.internal.csr'.
information/base: Writing private key to '/appl/icinga/icinga2/var/data/certs//avl4119p.it.internal.key'.
information/base: Writing certificate signing request to '/appl/icinga/icinga2/var/data/certs//avl4119p.it.internal.csr'.
information/cli: Signing CSR with CA and writing certificate to '/appl/icinga/icinga2/var/data/certs//avl4119p.it.internal.crt'.
information/pki: Writing certificate to file '/appl/icinga/icinga2/var/data/certs//avl4119p.it.internal.crt'.
information/cli: Copying CA certificate to '/appl/icinga/icinga2/var/data/certs//ca.crt'.
information/cli: API user config file '/etc/icinga2/conf.d/api-users.conf' already exists, not creating config file.
information/cli: Enabling the 'api' feature.
information/cli: Updating 'NodeName' constant in '/etc/icinga2/constants.conf'.
information/cli: Backup file '/etc/icinga2/constants.conf.orig' already exists. Skipping backup.
information/cli: Updating 'ZoneName' constant in '/etc/icinga2/constants.conf'.
information/cli: Backup file '/etc/icinga2/constants.conf.orig' already exists. Skipping backup.
Done.

Now restart your Icinga 2 daemon to finish the installation!

---

there are still files used in default path /etc/icinga2/ by icinga2 when started with -D ConfDir
examples:
/etc/icinga2/icinga2.conf
/etc/icinga2/constants.conf
/etc/icinga2/zones.conf
/etc/icinga2/conf.d/api-users.conf
/etc/icinga2/features-avaiable/*
/etc/icinga2/features-enabled/*

---

[root@avl4119p icingaweb2]# icinga2 troubleshoot --version -D ConfDir=/appl/icinga/icinga2/etc/icinga2 -D LogDir=/appl/logs/icinga/icinga2 -D DataDir=/appl/icinga/icinga2/var/data -D CacheDir=/appl/icinga/icinga2/var/cache -D SpoolDir=/appl/icinga/icinga2/var/spool -D InitRunDir=/appl/icinga/icinga2/var/run
icinga2 - The Icinga 2 network monitoring daemon (version: r2.10.2-1)

Copyright (c) 2012-2018 Icinga Development Team (https://icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

System information:
  Platform: Oracle Linux Server
  Platform version: 7.6
  Kernel: Linux
  Kernel version: 4.1.12-124.24.1.el7uek.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

Application information:

General paths:
  **Config directory: /etc/icinga2**
  Data directory: /appl/icinga/icinga2/var/data
  Log directory: /appl/logs/icinga/icinga2
  Cache directory: /appl/icinga/icinga2/var/cache
  Spool directory: /appl/icinga/icinga2/var/spool
  Run directory: /appl/icinga/icinga2/var/run

Old paths (deprecated):
  Installation root: /usr
  Sysconf directory: /etc
  Run directory (base): /run
  Local state directory: /var

Internal paths:
  Package data directory: /usr/share/icinga2
  State path: /appl/icinga/icinga2/var/data/icinga2.state
  Modified attributes path: /appl/icinga/icinga2/var/data/modified-attributes.conf
  Objects path: /appl/icinga/icinga2/var/cache/icinga2.debug
  Vars path: /appl/icinga/icinga2/var/cache/icinga2.vars
  PID path: /appl/icinga/icinga2/var/run/icinga2.pid
[root@avl4119p icingaweb2]#

----
ZonesDir related issue:

/usr/sbin/icinga2 console -D ConfigDir=/appl/icinga/icinga2/etc/icinga2 -D ConfDir=/appl/icinga/icinga2/etc/icinga2 -D LogDir=/appl/logs/icinga/icinga2 -D DataDir=/appl/icinga/icinga2/var/data -D CacheDir=/appl/icinga/icinga2/var/cache -D SpoolDir=/appl/icinga/icinga2/var/spool -D InitRunDir=/appl/icinga/icinga2/var/run
Icinga 2 (version: r2.10.2-1)
Type $help to view available commands.
<1> => ZonesDir
"/etc/icinga2/zones.d"
<2> => ^C
[icinga@avl4119p conf.d]$ /usr/sbin/icinga2 console -D ConfigDir=/appl/icinga/icinga2/etc/icinga2 -D ConfDir=/appl/icinga/icinga2/etc/icinga2 -D LogDir=/appl/logs/icinga/icinga2 -D DataDir=/appl/icinga/icinga2/var/data -D CacheDir=/appl/icinga/icinga2/var/cache -D SpoolDir=/appl/icinga/icinga2/var/spool -D InitRunDir=/appl/icinga/icinga2/var/run -D ZonesDir=/appl/icinga/icinga2/etc/icinga2/zones.d
Icinga 2 (version: r2.10.2-1)
Type $help to view available commands.
<1> => ZonesDir
"/appl/icinga/icinga2/etc/icinga2/zones.d"
<2> => ^C

## Possible Solution

ensure that all conf files and feature related files are considering ConfDir

## Steps to Reproduce (for bugs)

see under Current Behavior

## Your Environment

* Version used (`icinga2 --version`): r2.10.2-1
* Operating System and version: Oracle Linux Server release 7.6
* Enabled features (`icinga2 feature list`): api checker ido-mysql mainlog notification
* Icinga Web 2 version and modules (System - About): 2.6.2
* Config validation (`icinga2 daemon -C`):

[root@avl4119p icingaweb2]# icinga2 daemon -C -D ConfDir=/appl/icinga/icinga2/etc/icinga2 -D LogDir=/appl/logs/icinga/icinga2 -D DataDir=/appl/icinga/icinga2/var/data -D CacheDir=/appl/icinga/icinga2/var/cache -D SpoolDir=/appl/icinga/icinga2/var/spool -D InitRunDir=/appl/icinga/icinga2/var/run
[2019-03-05 02:48:11 +0100] information/cli: Icinga application loader (version: r2.10.2-1)
[2019-03-05 02:48:11 +0100] information/cli: Loading configuration file(s).
[2019-03-05 02:48:11 +0100] information/ConfigItem: Committing config item(s).
[2019-03-05 02:48:11 +0100] information/ApiListener: My API identity: avl4119p.it.internal
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 ScheduledDowntime.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 11 Services.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 IcingaApplication.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 Host.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 FileLogger.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 2 NotificationCommands.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 12 Notifications.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 NotificationComponent.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 2 HostGroups.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 ApiListener.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 2 Downtimes.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 CheckerComponent.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 3 Zones.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 Endpoint.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 ApiUser.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 User.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 IdoMysqlConnection.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 215 CheckCommands.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 1 UserGroup.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 3 ServiceGroups.
[2019-03-05 02:48:11 +0100] information/ConfigItem: Instantiated 3 TimePeriods.
[2019-03-05 02:48:11 +0100] information/ScriptGlobal: Dumping variables to file '/appl/icinga/icinga2/var/cache/icinga2.vars'
[2019-03-05 02:48:11 +0100] information/cli: Finished validating the configuration file(s).
[root@avl4119p icingaweb2]#

* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.

**Comment by Al2Klimov** [comment]
Hello @abaumgaertner!

Please read the doc (you linked) carefully. Are you sure about "ConfDir"?

Best,
AK

**Comment by dnsmichi** [comment]
Explicitly set `ZonesDir` then as a workaround.

**Comment by Al2Klimov** [comment]
TL;DR: It's "Conf**ig**Dir".
